oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,165 @@
+
+2. Create a ``data.py`` file to house simple data objects that can be passed from your app's function to the calling app. By creating these objects, we can avoid both passing Django model objects or querysets directly and having to serialize data. Other apps may import data classes from ``data.py`` in additional to functionality from ``api.py``. See data.py_ for more details.
+
+For example:
+
+.. code-block:: python
+
+    from django.conf.settings import UNSUPPORTED_PROGRAM_UUIDS
+
+    from .data import ProgramData
+    from .models_api import get_programs as _get_programs
+
+    def get_supported_programs():
+        """
+        Gets all programs that aren't in UNSUPPORTED_PROGRAM_UUIDS settings
+        """
+        all_programs = _get_programs() # returns queryset
+        supported_programs [
+          ProgramData(
+            uuid=program.uuid,
+            title=program.title,
+            status=program.status
+          )
+          for program in all_programs
+          if program.uuid not in UNSUPPORTED_PROGRAM_UUIDS
+        ]
+        return supported_programs
+
+.. _data.py:
+data.py
++++++++
+This file should include the public data structures for the app that can be passed between apps without exposing internal features. These should be used instead of sending Django model objects or querysets to apps that call the functions in ``api.py``. This file should not import anything other than stdlib modules, so that it may be imported by any other app without issue. These data objects should be simple objects with all business logic handled by ``api.py``. They may however perform simple validation, as long as it is self-contained (doesn't reach out to database, network, or any code outside of the class)
+
+For example:
+
+.. code-block:: python
+
+
+    from dataclasses import dataclass
+    from enum import Enum
+
+    def ProgramStatus(Enum):
+        ACTIVE = "active"
+        RETIRED = "retired"
+
+    @dataclass
+    class ProgramData:
+        uuid: str
+        title: str
+
+.. _models_api.py:
+models_api.py
++++++++++++++
+.. Reviewer NOTE: I don't feel like I entirely understand models_api.py. My main issue is I don't understand the benefit of the additional layer versus api.py just calling ``Thing.objects.get()``.
+
+
+.. _rest_api:
+rest_api/
++++++++++
+Each app should self-contain it's related REST API, meaning it should live within the app and not in a separate "api" app where it would mingle with other apps' REST API code. Since the Python API is using ``api.py``, the REST API code (often Django Rest Framework) should live in a ``rest_api`` folder.
+
+.. Reviewer NOTE: Do we want to detail what a standard ``rest_api`` looks like? For example having a serializers.py, permissions.py, versioned folders of APIs, etc. I'm not sure how much of that is common DRF and how much is Open edX.
```

Comment:
Just calling this note out because it doesn't show up if you're reviewing the render

Reply:
I've thought about this some more and I don't know if I necessarily agree with the serializers-in-view suggestion. For example: Discovery has APIs for course, course runs, and programs. Without the a shared CourseRunSerializer, we would end up with so much duplicate code in the program serializer reimplementing the course and course run serializers. I realize it's a flexibility vs DRY argument, but I feel we have too many situations where the flexible option is infeasible. 

I really like the `services.py` and having everything call into that w/ dumb models and dumb APIs. I'm not sure how much I want to lean into a paradigm shift versus a better/consistent way to things we already do. Evolutionary vs revolutionary and all that.

Reply:
> I realize it's a flexibility vs DRY argument, but I feel we have too many situations where the flexible option is infeasible.

I think of it less as "DRY vs. Flexible", and more as "DRY vs. Decoupled". That may be quibbling semantics, but I think of the latter as more defensively minded. Whenever we serialize to something, we're making some kind of API contract (whether we really mean to or not). So when I think of a serializer in a view, I think, "This is how we're going to serialize this data _for this version of the REST API_, but adding a field to this serializer should not change the CSV we export in this report generation task." It's not that I necessarily _want_ the flexibility to name fields differently, but more that I want a new column suddenly showing up in the middle of my CSV file because a Serializer change meant for one place rippled out to another.

Generally speaking, we haven't been as disciplined as we should be about API compatibility.

Reply:
I think I was thinking the serializers would be versioned, but shared within the version. For example `/rest_api/v1/serializers.py` would contain all the serializers for v1 of our API. Then if you need to change or remove fields, you'd create a `/rest_api/v2` with it's own serializers and views. So it's a decouple at the version or at the versioned-view argument, and I don't feel strongly on that. 

I agree about the lack of API discipline, but `I don't want a new column suddenly showing up in the middle of my CSV` isn't version-relevant because we don't version bump on additions.

I added a reference to https://openedx.atlassian.net/wiki/spaces/AC/pages/18350757/edX+REST+API+Conventions since that what OEP-0013 points to as the source of truth. (I didn't know it existed and I don't know why we moved it from an OEP to a wiki page :shrug:)

Reply:
> I agree about the lack of API discipline, but "I don't want a new column suddenly showing up in the middle of my CSV" isn't version-relevant because we don't version bump on additions.

I'm not sure if I'm interpreting this statement correctly, but just to clarify my own point: We've declared that adding a field to a resource when doing a GET in REST is backwards compatible. But adding that same new field to a CSV may not be backwards compatible, especially if it bumps out other fields. Many CSV readers (including Python's default one) just think of CSV files as numbered columns, so code will read through it and just use numbered indexing to get the column for that data (there is `DictReader`, but my point is just that it's not universal to use named field lookups in scripts like it would be for REST). If I'm sharing the serializer between the two, then something that's a totally allowed and benign change for the REST API results in a contract break for the CSV file.

Maybe that's a contrived example. The main place I can realistically see serializers being shared are when there are linked objects or pushing stuff to the non-yet-implemented message queue. And maybe that's actually fine. I just feel more comfortable when we have more granular control and I know that making a change to make a contract change (even a minor, non-breaking one) is only ever done in an intentional manner.

That being said, I'm honestly fine with having a shared module starting out. We can always amend this if it starts to become a problem–it just means we'd do a little copying to decouple later, and that's not that big a deal.